File: gui/src/core/Parser.cpp
/*
** EPITECH PROJECT, 2025
** Zappy
** File description:
** The Parser class implementation
*/
/**
 * @file Parser.cpp
 * @brief The Parser class implementation
 * @author Christophe VANDEVOIR, Gianni TUERO, Lou PELLEGRINO,
 * Nicolas TORO, Olivier POUECH and Raphael LAUNAY
 */

#include "Parser.hpp"

namespace Gui
{
    Parser::Parser(int const &argc, char ** const &argv) : _argc(argc), _argv(argv)
    {
        _options = {
            {"h", "hostname", "\tmachine", "\tThe hostname of the server", HOSTNAME_OPTION, &Parser::_optionHostname},
            {"p", "port", "\tport", "\tThe port of the server", PORT_OPTION, &Parser::_optionPort},
            {"H", "help", "\t", "\tDisplay this help message", META_OPTION, &Parser::_optionHelp},
            {"V", "version", "\t", "\tDisplay the version of the program", META_OPTION, &Parser::_optionVersion},
            {"A", "authors", "\t", "\tDisplay the authors of the project", META_OPTION, &Parser::_optionAuthors},
        };
        std::string arg = "";
        bool found = false;

        DEBUG << "Parsing command line arguments";
        for (size_t index = 1; index < static_cast<size_t>(_argc); index++) {
            arg = _argv[index];
            found = false;
            for (const auto &option : _options) {
                if (arg == "-" + option.shortName || arg == "--" + option.longName) {
                    (this->*(option.function))(index);
                    found = true;
                    break;
                }
            }
            if (!found) {
                throw Lib::Exceptions::Critical("Invalid option: \"" + arg + "\""
                    "\nPlease use -H or --help to see the list of available options.\"");
            }
        }
        for (const auto &option : _options) {
            if (option.type != NOT_REQUIRED && option.type != META_OPTION
            && (_optionsFound & option.type) == 0) {
                throw Lib::Exceptions::Critical("Missing option: -"
                    + option.shortName + " or --" + option.longName);
            }
        }
    }

    Parser::~Parser()
    {
        DEBUG << "Destroying Parser";
    }

    const std::string &Parser::getHostname() const
    {
        return _hostname;
    }

    const int &Parser::getPort() const
    {
        return _port;
    }

    void Parser::_optionHostname(size_t &index)
    {
        if (index + 1 >= static_cast<size_t>(_argc)) {
            throw Lib::Exceptions::Critical("Missing argument for -h (or --hostname) option");
        }
        _hostname = _argv[++index];
        if (_hostname.empty() || std::count(_hostname.begin(), _hostname.end(), '.') != 3) {
            throw Lib::Exceptions::Critical("Invalid argument for -h (or --hostname) option, it should be a valid hostname or IP address");
        }
        _optionsFound |= HOSTNAME_OPTION;
    }

    void Parser::_optionPort(size_t &index)
    {
        if (index + 1 >= static_cast<size_t>(_argc)) {
            throw Lib::Exceptions::Critical("Missing argument for -p or --port option");
        }
        try {
            _port = std::stoi(_argv[++index]);
        } catch (const std::invalid_argument &) {
            throw Lib::Exceptions::Critical("Invalid argument for -p (or --port) option, it must be a number between 0 and " + std::to_string(MAX_PORT_NUMBER));
        }
        if (_port < 0 || _port > MAX_PORT_NUMBER) {
            throw Lib::Exceptions::Critical("Invalid argument for -p (or --port) option, it must be a number between 0 and " + std::to_string(MAX_PORT_NUMBER));
        }
        _optionsFound |= PORT_OPTION;
    }

    void Parser::_optionHelp(size_t &)
    {
        std::cout << BOLD "USAGE:" RESET << std::endl
            << "\t" << _argv[0] << " -h hostname -p port" << std::endl << std::endl
            << BOLD "OPTIONS:" RESET << std::endl;
        for (const auto &option : _options) {
            if (option.type != META_OPTION) {
                std::cout << "\t-" << option.shortName << ", --" << option.longName
                    << "\t" << option.argument << "\t\t" << option.description << std::endl;
            }
        }
        std::cout << std::endl << BOLD "META-OPTIONS:" RESET << std::endl;
        for (const auto &option : _options) {
            if (option.type == META_OPTION) {
                std::cout << "\t-" << option.shortName << ", --" << option.longName
                    << "\t" << option.argument << "\t\t" << option.description << std::endl;
            }
        }
        throw Lib::Exceptions::Success("");
    }

    void Parser::_optionVersion(size_t &)
    {
        std::cout << BOLD "VERSION:" RESET << std::endl
            << "\tZappy GUI version 1.0.0" << std::endl
            << "\tBuilt on: " __DATE__ " at " __TIME__ << std::endl;
        throw Lib::Exceptions::Success("");
    }

    void Parser::_optionAuthors(size_t &)
    {
        std::cout << BOLD "AUTHORS:" RESET << std::endl
            << ITALIC "\t- Students of EPITECH Nice 2025 -" RESET << std::endl
            << "\tChristophe VANDEVOIR" << std::endl
            << "\tGianni TUERO" << std::endl
            << "\tLou PELLEGRINO" << std::endl
            << "\tNicolas TORO" << std::endl
            << "\tOlivier POUECH" << std::endl
            << "\tRaphaël LAUNAY" << std::endl;
        throw Lib::Exceptions::Success("");
    }
}

File: bonus/ai/parser.py
import sys
from exceptions import *



class OptionType:
    NOT_REQUIRED = 0
    HOSTNAME_OPTION = 1 << 0
    PORT_OPTION = 1 << 1
    NAME_OPTION = 1 << 2
    META_OPTION = 1 << 3



class Option:
    def __init__(self, short_name, long_name, argument, description, option_type, function):
        self.short_name = short_name
        self.long_name = long_name
        self.argument = argument
        self.description = description
        self.type = option_type
        self.function = function



class Parser:
    MAX_PORT_NUMBER = 65535


    def __init__(self, argv):
        self._argv = argv
        self._options = [
            Option("h", "hostname", "machine", "The hostname of the server", OptionType.HOSTNAME_OPTION, self.__option_hostname),
            Option("p", "port", "port", "The port of the server", OptionType.PORT_OPTION, self.__option_port),
            Option("n", "name", "name", "The name of the team", OptionType.NAME_OPTION, self.__option_name),
            Option("c", "child", "", "The AI is a child", OptionType.NOT_REQUIRED, self.__option_child),
            Option("H", "help", "", "Display this help message", OptionType.META_OPTION, self.__option_help),
            Option("V", "version", "", "Display the version of the program", OptionType.META_OPTION, self.__option_version),
            Option("A", "authors", "", "Display the authors of the project", OptionType.META_OPTION, self.__option_authors)
        ]
        self._options_found = 0
        self._hostname = ""
        self._port = 0
        self._name = ""
        self._child = False

        self.__parse()


    def __parse(self):
        index = 0
        while index < len(self._argv):
            arg = self._argv[index]
            found = False
            for option in self._options:
                if arg == f"-{option.short_name}" or arg == f"--{option.long_name}":
                    index = option.function(index)
                    found = True
                    break
            if not found:
                raise CriticalException(
                    f"Invalid option: \"{arg}\"\nPlease use -H or --help to see the list of available options.")

        for option in self._options:
            if option.type != OptionType.NOT_REQUIRED and option.type != OptionType.META_OPTION:
                if (self._options_found & option.type) == 0:
                    raise CriticalException(
                        f"Missing option: -{option.short_name} or --{option.long_name}")


    def get_hostname(self):
        return self._hostname


    def get_port(self):
        return self._port


    def get_name(self):
        return self._name


    def get_child(self):
        return self._child


    def __option_hostname(self, index):
        if index + 1 >= len(self._argv):
            raise CriticalException(
                "Missing argument for -h (or --hostname) option")

        self._hostname = self._argv[index + 1]
        if not self._hostname or self._hostname.count('.') != 3:
            raise CriticalException(
                "Invalid argument for -h (or --hostname) option, it should be a valid hostname or IP address")

        self._options_found |= OptionType.HOSTNAME_OPTION
        return index + 2


    def __option_port(self, index):
        if index + 1 >= len(self._argv):
            raise CriticalException("Missing argument for -p or --port option")

        try:
            self._port = int(self._argv[index + 1])
        except ValueError:
            raise CriticalException(
                f"Invalid argument for -p (or --port) option, it must be a number between 0 and {self.MAX_PORT_NUMBER}")

        if self._port < 0 or self._port > self.MAX_PORT_NUMBER:
            raise CriticalException(
                f"Invalid argument for -p (or --port) option, it must be a number between 0 and {self.MAX_PORT_NUMBER}")

        self._options_found |= OptionType.PORT_OPTION
        return index + 2


    def __option_name(self, index):
        if index + 1 >= len(self._argv):
            raise CriticalException(
                "Missing argument for -n (or --name) option")

        self._name = self._argv[index + 1]
        if not self._name:
            raise CriticalException(
                "Invalid argument for -n (or --name) option")

        self._options_found |= OptionType.NAME_OPTION
        return index + 2


    def __option_child(self, index):
        self._child = True
        return index + 1


    def __option_help(self, index):
        print("\033[1mUSAGE:\033[0m")
        print(f"\t{sys.argv[0]} -h hostname -p port -n name\n")
        print("\033[1mOPTIONS:\033[0m")

        for option in self._options:
            if option.type != OptionType.META_OPTION:
                print(
                    f"\t-{option.short_name}, --{option.long_name}\t{option.argument}\t\t{option.description}")

        print("\n\033[1mMETA-OPTIONS:\033[0m")
        for option in self._options:
            if option.type == OptionType.META_OPTION:
                print(
                    f"\t-{option.short_name}, --{option.long_name}\t{option.argument}\t\t{option.description}")

        raise SuccessException()


    def __option_version(self, index):
        print("\033[1mVERSION:\033[0m")
        print("\tZappy AI version 1.0.0")
        #print(f"\tBuilt on: {__date__} at {__time__}")
        raise SuccessException()


    def __option_authors(self, index):
        print("\033[1mAUTHORS:\033[0m")
        print("\t\033[3m- Students of EPITECH Nice 2025 -\033[0m")
        print("\tChristophe VANDEVOIR")
        print("\tGianni TUERO")
        print("\tLou PELLEGRINO")
        print("\tNicolas TORO")
        print("\tOlivier POUECH")
        print("\tRaphaël LAUNAY")
        raise SuccessException()
